Fixes ncc compiler to generate the right bundle for all environments

ajainarayanan opened the pull request:
- ~Post nodejs upgrade ncc compiler seems to behave in a completely different way~
  - ~Not modify the CDAP UI static assets based on the compilation step~
  - ~Not copy right static asset folders post compilation.~
- ~This forced us to remove ncc compiler for now to unblock SDK usage.~
- ~Need to find out what we can do compile our node app by ourselves. Related JIRA - [CDAP-15760](https://issues.cask.co/browse/CDAP-15760)~

**UPDATE**
- Upgrades `ncc` compiler
- Modifies the location of all dist folders to be consistent instead of it being relative from current directory.
- Introduces a top level `public` directory that now contains all the dist folders under `cdap-ui/server/` folder. This is here because this is where `express` is running. It makes sense to be there since express is serving all these static files
- Modifies webpack config and Gulpfiles to account for the public folder under server directory.

elfenheart commented on cdap-ui/pom.xml:
```diff
@@ -260,24 +260,6 @@
                   <arguments>run distribute</arguments>
                 </configuration>
               </execution>
-              <execution>
```
wouldn't this prevent our build to run the unit test?

ajainarayanan replied:
Yes. Right now that unit test is seldom being updated or maintained. Since we now have separate build plans let's skip on running unit tests during build. If you feel the tests add value let me know.

elfenheart commented on cdap-ui/yarn.lock:
```diff
@@ -3620,7 +3590,7 @@ babel-traverse@^6.23.1, babel-traverse@^6.26.0:
     lodash "^4.17.4"
 
-babel-types@^6.23.0, babel-types@^6.24.1, babel-types@^6.26.0:
+babel-types@^6.23.0, babel-types@^6.26.0:
```
was there a change to `package.json`?

ajainarayanan replied:
Yeah we removed the ncc module right?

elfenheart replied:
my bad, for some reason the package.json changes was not showing up

elfenheart commented:
@ajainarayanan I think we'll have to add the yarn production install. Right now the entire node_modules are being copied

ajainarayanan committed:
[CDAP-15760] Fixes ncc compiler to generate the right bundle for all environments

  - Upgrades ncc compiler node module
  - Modifies webpack config based on new public containing all dist folders
  - Modifies Gulpfile  based on new public containing all dist folders
  - Fixes express based on new public dist folders + modifies graphql based on new ncc compiler upgrade
  - Adds server/public to .gitignore
  - Updates pom.xml to copy right artifacts based on new public dist folders
  - Adds new shared dll modules
  - Fixes sass variable syntax

elfenheart commented:
Need to update the lockfile. Build is failing @ajainarayanan

elfenheart commented:
After updating the yarn.lock, UI does not start
![Screen Shot 2019-08-19 at 11 22 17 AM](https://user-images.githubusercontent.com/4398643/63289334-b2e6ab80-c273-11e9-936c-ed73f9289120.png)

ajainarayanan commented:
@elfenheart I have updated the yarn.lock file. ~Are you sure you restarted the node server after the build? I am able to build and the UI shows up fine though.~

This happened because of incorrect configuration of `.gitignore`. Have reset it to not ignore dll files and have checked in the dlls

elfenheart reviewed (approved):
👍 LGTM

ajainarayanan commented:
Latest green build: https://builds.cask.co/browse/CDAP-UDUT390-2
